docs/proposals/initializers.md

```diff
@@ -0,0 +1,110 @@
+
+## REST API Changes
+
+All API objects will support a new endpoint as a sub-resource that allows non-initialized objects to be edited via a PUT that does not carry the same validation constraints as normal updates of a resource.
+
+For example:
+
+```PUT /api/{version}/{group}/{kind}/{name}/initialize/{initializer-name}```
+
+The input to the request would be the normal serialization of the object.  The input object would be persisted, and the named *initializer* would be removed from the object's set of **ObjectMeta.Initializers**.  The only validation that must pass on the object is the validation of the *ObjectMeta* related data.  If the named object was already initialized, the API server would error.  If the initializer was not in the list of **ObjectMeta.Initializers**, the request would error.
+
+In addition, each object would support an API to mark the object as completely initialized.
+
+For example:
+
+```POST /api/initialize```
+
+The input to the request would be a *Initialize* object, that would work similar to *Scale* with HPA that would take a subresource reference to know the object that is to be transitioned from non-initialized to initialized.  If the object was already initialized,
+the API would error.  If the object failed strict validation, the system would error.
+
+**NOTE: I will go back and fill in more details here**
+
+## Making the transition to initialized
+
+A new controller is introduced into the system that looks for objects that have no more named "initializers" but are not yet "initialized".
+
+It introspects the set of registered *Initializers*, and uses the *Discovery API* and a generic API client to know what resource types to WATCH that are not yet initialized.  It then creates a set of watchers that look for objects whose list of "initializers" is empty, but the object is not yet "initialized".
+
+It then performs an "/api/initialize" action for each resource that meets the required criteria.  If the API action succeeded, the object is now initialized for use by the system.  If the API action fails, it's considered to be in an error state that is unrecoverable.  When this occurs, the controller will permanently delete the resource from the API server and log an event because the object would forever fail strict validation.
+
+## Existing controller manager components
+
+Objects like a ReplicationController should treat a non-initialized pod as a pod, meaning it should not create new ones to satisfy its desired replica set.
+
+Objects like the Scheduler should treat a non-initialized pod as a pod not yet ready to be scheduled.
```

Comment:
What about get?  Probably has a similar impact if a naive client is polling
for the existence of a specific service by name (kube-dns) in a loop - when
we introduce this path, that client is broken because the service isn't
initialized yet.

On Thu, Jan 7, 2016 at 9:33 PM, Brian Grant <notifications@github.com>
wrote:

> In docs/proposals/initializers.md
> <https://github.com/kubernetes/kubernetes/pull/17305#discussion_r49153363>
> :
>
> > +
> > +**NOTE: I will go back and fill in more details here**
> > +
> > +## Making the transition to initialized
> > +
> > +A new controller is introduced into the system that looks for objects that have no more named "initializers" but are not yet "initialized".
> > +
> > +It introspects the set of registered *Initializers*, and uses the *Discovery API* and a generic API client to know what resource types to WATCH that are not yet initialized.  It then creates a set of watchers that look for objects whose list of "initializers" is empty, but the object is not yet "initialized".
> > +
> > +It then performs an "/api/initialize" action for each resource that meets the required criteria.  If the API action succeeded, the object is now initialized for use by the system.  If the API action fails, it's considered to be in an error state that is unrecoverable.  When this occurs, the controller will permanently delete the resource from the API server and log an event because the object would forever fail strict validation.
> > +
> > +## Existing controller manager components
> > +
> > +Objects like a ReplicationController should treat a non-initialized pod as a pod, meaning it should not create new ones to satisfy its desired replica set.
> > +
> > +Objects like the Scheduler should treat a non-initialized pod as a pod not yet ready to be scheduled.
>
> The proposal was to not return not-fully-initialized resources in
> list/watch by default, with a query param to request all/uninitialized.
>
> A longer-term way to deal with pre-initialization and post-deletion could
> be to dissociate the resource name from the resource, by storing the
> resource by uid.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/17305/files#r49153363>.
>